Refactor with Value Types

nicocharlery opened the pull request:
Based on our discussion, we will follow the rules below:

1. A Builder pattern or a DSL are high-level interfaces that can be added later, if need be.
2. At a low level, we want to pass a FileHeader type 
3. Use Elixir primitives (like `Date` and `Time`)
4. Have types upfront in charge or coercion/cast, validation and documentation
```elixir
{:ok, t} | {:error, ...} = AccountNumber.new("12345")
```
This will give us also the opportunity to document
5. Validation can then be handled without a library like `Vex`, but with bare Elixir modules (eg. `Enum.count/1` or `String.length`).

*Reference:* https://github.com/hollar/ex_ach/pull/2#issuecomment-370916498

nicocharlery committed:
Update README

Commit:
Use new way to structure header

nicocharlery committed:
Delete previous file header implementation

nicocharlery committed:
Use new implementation of batch header

nicocharlery committed:
Delete previous batch header implementation

nicocharlery committed:
Add fields validations

nicocharlery committed:
Accumulate field errors

When two errors are present in a given value, we want to return the set of errors at once

nicocharlery committed:
Validate company_identification

nicocharlery committed:
Lint code

Ran `mix format`

nicocharlery committed:
Complete batch header validation

Commit:
Extract Field module

We can then test the validator independently and avoid testing the description of them in the fields.
If we need to test the inclusion of a validation rule, we can use the `specifications` public function.

Commit:
Use ExAch.Field validator

nicocharlery committed:
Add FieldValidator test

nicocharlery committed:
Test Field integration with the validator

Commit:
Use field_validator to validate file header fields

nicocharlery committed:
Remote Field assertions

We have a test for Field and for the validator.
A wrong data in the test would actually trigger an error.

Commit:
Add batch entry fields

Commit:
Generate File creation date and time

As indicated in ACH file specification FileCreationDate and
FileCreationTime are used to indicate when ach file is being generated.
In this commit we are generating those values internally, rather then
allowing user of the library pass in those values.

Commit:
Add strict_alpha_num_string type

Rename type :string into :alpha_num_string, as name for type string
indicates that it can be any valid string, as looking at the
implementation it looks like it's an alpha numeric string

Commit:
Add specs and docs

nicocharlery committed:
Lint

Run `mix format`

nicocharlery committed:
Format and add missing types definition in fields

nicocharlery commented:
@nicholasjhenry @frahugo  This PR is ready for review. As it's quite lengthy, I can walk you through the changes made, or if not, I'd suggest you to browse the project as a whole as it's a complete rewrite of the previous implementation.

Bokay committed:
Merge pull request #4 from hollar/feature/generate-batch-entry-details

Generate batch entry details records

frahugo commented on README.md:
````diff
@@ -20,45 +20,124 @@ and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
 ## Usage
+Create a file header:
+```elixir
+alias ExAch.File.Header
````
Should be `ExAch.File`

frahugo commented on README.md:
````diff
@@ -20,45 +20,124 @@ and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
 ```elixir
-iex> ach = ExAch.new()
-%ExAch.Ach{}
+alias ExAch.Batch.Header
````
Should be `ExAch.Batch`

frahugo commented on README.md:
```diff
@@ -20,45 +20,124 @@ and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
+{:ok, identification_number} = Batch.Entry.Fields.IdentificationNumber.new(1000)
+
+{:ok, batch_entry} =
+  ExAch.Batch.Entry.new(
```
remove `ExAch`

frahugo commented on README.md:
```diff
@@ -20,45 +20,124 @@ and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
+{:ok, company_discretionary_data} = Batch.Header.Fields.CompanyDiscretionaryData.new("Data1")
+
+{:ok, batch_header} =
+  ExAch.Batch.Header.new(
```
remove `ExAch.`

frahugo commented on README.md:
```diff
@@ -20,45 +20,124 @@ and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
+{:ok, file_creation_time} = File.Header.Fields.FileCreationTime.new(~T[23:00:07.000])
+
+{:ok, file_header} =
+  ExAch.File.Header.new(
```
remove `ExAch.`

frahugo commented on README.md:
````diff
@@ -20,45 +20,124 @@ and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
+```elixir
+alias ExAch.Batch
+
+{:ok, batch} = ExAch.Batch.new(header, entries)
````
Should be `ExAch.Batch.new(batch_header, entries)`

frahugo commented on README.md:
```diff
@@ -20,45 +20,124 @@ and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
-{:error, [{:error, :immediate_destination, :presence, "must be present"}]}
+alias ExAch.{File, Batch}
+
+{:ok, ach} = ExAch.File.new(file_header, batches)
```
Maybe this instead?

```
{:ok, ach} = ExAch.File.new(file_header, [batch])
```

frahugo commented on README.md:
```diff
@@ -20,45 +20,124 @@ and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
+alias ExAch.{File, Batch}
+
+{:ok, ach} = ExAch.File.new(file_header, batches)
+%ExAch.File{file_header: %FileHeader{}, batches: [%Batch]}
```
Maybe we can get rid of this line

frahugo commented on README.md:
````diff
@@ -20,45 +20,124 @@ and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
 ```
 
+
+ach = ExAch.new # => %ExAch.Ach{}
````
Are the next lines needed?

frahugo commented on lib/ex_ach/batch/header/fields/standard_entry_class_code.ex:
```diff
@@ -0,0 +1,13 @@
+  """
+
+  @type t :: %__MODULE__{}
+  use ExAch.Field, specifications: [type: :atom, inclusion: [:web, :ccd, :ppd, :ctx, :tel]]
```
I wonder if we should only include the codes the library is supporting.  Or we raise somewhere a "NotImplemented" exception.

frahugo commented on lib/ex_ach/field_validator.ex:
```diff
@@ -0,0 +1,89 @@
+
+  defp do_validate(content, {field_name, :type, :alpha_num_string}) when is_binary(content) do
+    if !Regex.match?(~r/^[0-9A-Za-z ]+$/, content) do
+      {field_name, :type, "Must be alphanum"}
```
Maybe this? `Must be letters or digits`

frahugo commented on lib/ex_ach/field_validator.ex:
```diff
@@ -0,0 +1,89 @@
+  def validate(content, validations) do
+    validations
+    |> Enum.map(&do_validate(content, &1))
+    |> Enum.reject(fn value -> value == nil end)
```
`Enum.reject(&is_nil/1)`

frahugo commented on test/ex_ach/batch/entry_test.exs:
```diff
@@ -0,0 +1,66 @@
+      assert batch_entry.discretionary_data == discretionary_data
+
+      # Mandatory fields generated by library
+      assert batch_entry.record_type_code == %RecordTypeCode{content: 6}
```
It might be better to hide the implementation of the content (maybe?):

`assert batch_entry.record_type_code == %RecordTypeCode.new(6)`

frahugo commented on test/ex_ach/batch/header_validation_test.exs:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  describe "validating service_class_code" do
+    test "a valid value returns successfully" do
```
Remove `a ` and `an `

frahugo commented on test/ex_ach/field_test.exs:
```diff
@@ -0,0 +1,25 @@
+  end
+
+  setup_all do
+    :dummy_field
```
I think we could remove this line by extracting the field_name logic into a `field_name` function in `Field`, and then add a test here: `assert Field.field_name(dummy) == :dummy_field`.

frahugo commented on test/ex_ach/field_test.exs:
```diff
@@ -0,0 +1,25 @@
+    :ok
+  end
+
+  describe "validating content" do
```
Missing a test in the `describe` block

frahugo commented on test/ex_ach/field_validator_test.exs:
```diff
@@ -0,0 +1,174 @@
+    test "a non alpha numeric string type returns an error" do
+      specifications = [{:field_name, :type, :alpha_num_string}]
+      errors = FieldValidator.validate(123, specifications)
+
```
Remove blank lines.  When the test has only 3 lines (or less), they should be grouped.

frahugo commented on test/ex_ach/field_validator_test.exs:
```diff
@@ -0,0 +1,174 @@
+      specifications = [{:field_name, :max_length, 3}]
+      errors = FieldValidator.validate("1234", specifications)
+
+      assert errors == [{:field_name, :max_length, "Must be less than 3 character(s)"}]
```
The message should be "Must be less than or equal to 3 character(s)"

frahugo commented on test/ex_ach/field_validator_test.exs:
```diff
@@ -0,0 +1,174 @@
+      assert errors == [{:field_name, :length, "Must be 3 character(s)"}]
+    end
+
+    test "a integer too long returns an error" do
```
Missing a test with less characters passed.

frahugo commented on test/ex_ach/file/header_test.exs:
```diff
@@ -0,0 +1,63 @@
+      assert file_header.format_code == %FormatCode{content: 1}
+
+      # Optional fields passed by user
+      assert file_header.file_creation_time
```
`assert %FileCreationTime{} = file_header.file_creation_time`

frahugo commented on test/ex_ach/file/header_test.exs:
```diff
@@ -0,0 +1,63 @@
+      assert file_header.blocking_factor == %BlockingFactor{content: 10}
+      assert file_header.format_code == %FormatCode{content: 1}
+
+      # Optional fields passed by user
```
Wrong comment

frahugo commented on test/ex_ach/file/header_test.exs:
```diff
@@ -0,0 +1,63 @@
+      # Optional fields passed by user
+      assert file_header.file_creation_time
+
+      assert file_header.immediate_destination_name == %ImmediateDestinationName{
```
Use the `new` function to hide content implementation.

frahugo reviewed (changes requested):
Excellent job!  Few minor things to fix.  We can do that together.

nicholasjhenry commented on lib/ex_ach/batch/entry/fields/dfi_account_number.ex:
```diff
@@ -0,0 +1,12 @@
+  @type t :: %__MODULE__{}
+  use ExAch.Field,
+    specifications: [
+      type: :alpha_num_string,
```
@frahugo / @bokay seems like `alpha_numeric` would be an appropriate `type` name here.

Commit:
Add doctest for ExAch module

This doctest will serve as an integration test for the whole library and as a usage example.
Run:
`mix docs
open docs/index.html`

Commit:
Restrict entry standard entry class code list to CCD

Commit:
Change strict_alpha_num_string type to file_id_modifier

Commit:
Lint

Commit:
Hide Field implementation

Commit:
Lint

mix format

Commit:
Remove article in docstrings

Commit:
Force existing atom in Field test

Commit:
Add missing doctest

Commit:
Follow our style guide

Commit:
Adjust error message for max_length validation

Commit:
Fix length validation on integers

Commit:
Lint

Commit:
Rename alpha_num_string type to alpha_numeric

Commit:
Fix warning

Commit:
Rename default_content to value

Commit:
Move typespec t definition in Field

Commit:
Rename specifications to validation/validation_rules

Commit:
Change file_id_modifier type to a format validation

Using a format validation results in a more extensible library.

Commit:
Handle format validation with a given message

Commit:
Inline field validation

Commit:
Document record options
